core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/ClusterUpgradeImageFilter.java

```diff
@@ -109,9 +116,11 @@ private boolean permitLockedComponentsUpgrade(Image currentImage, Image image) {
         return currentImagePackageVersions.equals(imagePackageVersions);
     }
 
-    private boolean permitCmAndStackUpgrade(Image currentImage, Image image, String key) {
-        return upgradePermissionProvider.permitCmAndStackUpgrade(currentImage.getPackageVersions().get(key),
-                image.getPackageVersions().get(key));
+    private boolean permitCmAndStackUpgrade(Image currentImage, Image image, String versionKey, String buildNumberKey) {
+        String currentVersion = currentImage.getPackageVersions().get(versionKey);
+        String newVersion = image.getPackageVersions().get(versionKey);
+        return upgradePermissionProvider.permitCmAndStackUpgrade(currentVersion, newVersion)
+                || (currentVersion.equals(newVersion) && componentBuildNumberComparator.compare(currentImage, image, buildNumberKey));
```

Comment:
logically, this should be part of permitCmAndStackUpgrade method otherwise, it might be misleading, couldn't it?